File: agent/drand_dice.py
"""drand_dice.py — derive backgammon dice from a drand round digest.

Each turn's dice are `keccak256(drand_round_digest || turn_index_be8)
mod 36`, then unpacked into a `(d1, d2)` pair in the range 1..6.

This is the deterministic randomness primitive the README and the
per-turn KeeperHub sequence diagram refer to: anyone replaying a match
can re-fetch the same drand round, re-hash, and recover the same dice
without trusting the server.

Why drand vs. commit-reveal:
  - drand rounds are publicly attested by the League of Entropy and
    timestamped — there's no per-match coordination overhead.
  - The dice for round R cannot be predicted before R is published, and
    cannot be denied or forged after.
  - Commit-reveal still works as a fallback for offline / out-of-band
    matches; drand is just the cleaner primary path.

This module does NOT call out to a drand HTTP endpoint — fetching is the
caller's responsibility (KeeperHub workflow step in production, a test
fixture in unit tests). The helpers here only consume an already-fetched
round digest.
"""
from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha3_256


# Number of distinct unordered dice rolls = 6 * 6 = 36.
DICE_FACE_COUNT = 6
DICE_PAIR_COUNT = DICE_FACE_COUNT * DICE_FACE_COUNT


@dataclass(frozen=True)
class DiceRoll:
    """One backgammon dice roll. `d1` and `d2` are independent 1..6 faces;
    doubles are detected with `d1 == d2` (a backgammon double-rolls move).

    `round_number` and `turn_index` round-trip the inputs so a match
    record can carry them alongside the resulting `(d1, d2)` for replay
    verification."""
    d1: int
    d2: int
    round_number: int
    turn_index: int

    def __post_init__(self) -> None:
        if not (1 <= self.d1 <= DICE_FACE_COUNT):
            raise ValueError(f"d1 out of range: {self.d1}")
        if not (1 <= self.d2 <= DICE_FACE_COUNT):
            raise ValueError(f"d2 out of range: {self.d2}")

    @property
    def is_double(self) -> bool:
        return self.d1 == self.d2

    def as_tuple(self) -> tuple[int, int]:
        return (self.d1, self.d2)


def _keccak256(*chunks: bytes) -> bytes:
    """SHA3-256 (== keccak256 by spec) of the concatenation of chunks.

    Note: Python's hashlib.sha3_256 implements the FIPS-202 SHA3-256,
    which matches Solidity's `keccak256` — both are Keccak-f[1600] with
    the same rate/capacity and standard padding. So `_keccak256(b)` here
    produces the same 32 bytes as `keccak256(b)` on Sepolia.
    """
    h = sha3_256()
    for chunk in chunks:
        h.update(chunk)
    return h.digest()


def derive_dice(round_digest: bytes, turn_index: int, *,
                round_number: int | None = None) -> DiceRoll:
    """Derive a `(d1, d2)` roll from `round_digest` (the bytes returned
    by drand for round R) and `turn_index` (a per-match counter starting
    at 0).

    Encoding:
      bucket = keccak256(round_digest || turn_index_be8) mod 36
      d1 = bucket // 6 + 1     # 1..6
      d2 = bucket %  6 + 1     # 1..6

    `round_number` is preserved on the returned `DiceRoll` for replay
    verification but is NOT mixed into the hash — the round digest is
    already round-bound, so re-hashing the round number would only
    duplicate that binding.
    """
    if not isinstance(round_digest, (bytes, bytearray)):
        raise TypeError("round_digest must be bytes")
    if turn_index < 0:
        raise ValueError(f"turn_index must be non-negative, got {turn_index}")

    digest = _keccak256(bytes(round_digest), turn_index.to_bytes(8, "big"))
    bucket = int.from_bytes(digest, "big") % DICE_PAIR_COUNT
    d1 = bucket // DICE_FACE_COUNT + 1
    d2 = bucket % DICE_FACE_COUNT + 1
    return DiceRoll(
        d1=d1, d2=d2,
        round_number=round_number if round_number is not None else -1,
        turn_index=turn_index,
    )


def derive_dice_sequence(round_digest: bytes, n_turns: int, *,
                         round_number: int | None = None) -> list[DiceRoll]:
    """Convenience: derive dice for turns 0..n_turns-1 from one round
    digest. Used in tests; production code pulls a fresh drand round per
    turn so KeeperHub can interleave validation between them."""
    if n_turns < 0:
        raise ValueError(f"n_turns must be non-negative, got {n_turns}")
    return [derive_dice(round_digest, t, round_number=round_number)
            for t in range(n_turns)]

File: frontend/lib/drand_dice.ts
// drand_dice.ts — verifiable backgammon dice from a drand round.
//
// TypeScript port of `agent/drand_dice.py`. Each turn's dice are
//   bucket = keccak256(round_digest ‖ turn_index_be8) mod 36
//   d1 = bucket // 6 + 1   d2 = bucket % 6 + 1
// keccak256 is FIPS-202 SHA3-256 by spec, matching the Python (hashlib.sha3_256)
// and Solidity, so a roll derived here re-derives identically on any side.
//
// Why drand (not commit-reveal): a drand round is publicly attested by the
// League of Entropy and cannot be predicted before publication or forged after,
// so neither peer controls the roll and either can verify it — without a server
// or any per-match coordination. For human-vs-human each browser fetches the
// SAME round and derives identical dice; the round is a public decentralized
// beacon, so it stays serverless / off-chain like the Nostr relays.
//
// NOTE: both peers (and, for replay parity, the KeeperHub/Python path) must
// agree on which drand field is the "digest". We use the per-round `randomness`
// output here; keep `agent/drand_dice.py`'s caller aligned with the same field.

import { keccak256, type Hex } from "viem";

export const DICE_FACE_COUNT = 6;
export const DICE_PAIR_COUNT = DICE_FACE_COUNT * DICE_FACE_COUNT; // 36

export interface DiceRoll {
  d1: number;
  d2: number;
  roundNumber: number;
  turnIndex: number;
}

export function isDouble(roll: DiceRoll): boolean {
  return roll.d1 === roll.d2;
}

function hexToBytes(hex: Hex): Uint8Array {
  const s = hex.startsWith("0x") ? hex.slice(2) : hex;
  if (s.length % 2 !== 0) throw new Error(`odd-length hex: ${hex}`);
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(s.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function turnIndexBE8(turnIndex: number): Uint8Array {
  if (!Number.isInteger(turnIndex) || turnIndex < 0) {
    throw new Error(`turnIndex must be a non-negative integer, got ${turnIndex}`);
  }
  const buf = new Uint8Array(8);
  let n = BigInt(turnIndex);
  for (let i = 7; i >= 0; i--) {
    buf[i] = Number(n & 0xffn);
    n >>= 8n;
  }
  return buf;
}

/**
 * Derive a `(d1, d2)` roll from a drand round digest (hex, e.g. the round's
 * `randomness`) and a per-match `turnIndex` (counter from 0). Deterministic and
 * verifiable: anyone with the same round + turnIndex recovers the same dice.
 */
export function deriveDice(
  roundDigest: Hex,
  turnIndex: number,
  roundNumber = -1,
): DiceRoll {
  const digest = hexToBytes(roundDigest);
  const input = new Uint8Array(digest.length + 8);
  input.set(digest, 0);
  input.set(turnIndexBE8(turnIndex), digest.length);

  const bucket = Number(BigInt(keccak256(input)) % BigInt(DICE_PAIR_COUNT));
  return {
    d1: Math.floor(bucket / DICE_FACE_COUNT) + 1,
    d2: (bucket % DICE_FACE_COUNT) + 1,
    roundNumber,
    turnIndex,
  };
}

// ── drand HTTP fetch ────────────────────────────────────────────────────────
//
// Default to drand "quicknet" (3s rounds) via the public api.drand.sh mirror so
// per-turn roll latency is ~3s rather than mainnet's 30s. Both peers MUST use
// the same beacon (chain hash) for dice to agree. Verify the chain hash /
// endpoint against https://drand.love when wiring this up.
export const DRAND_QUICKNET_CHAIN_HASH =
  "52db9ba70e0cc0f6eaf7803dd07447a1f5477735fd3f661792ba94600c84e971";
const DRAND_BASE = `https://api.drand.sh/${DRAND_QUICKNET_CHAIN_HASH}`;

export interface DrandRound {
  round: number;
  randomness: Hex;
}

/** Fetch a drand round (defaults to the latest published round). */
export async function fetchDrandRound(round?: number): Promise<DrandRound> {
  const res = await fetch(`${DRAND_BASE}/public/${round ?? "latest"}`);
  if (!res.ok) throw new Error(`drand fetch failed: ${res.status}`);
  const j = (await res.json()) as { round: number; randomness: string };
  return { round: j.round, randomness: `0x${j.randomness}` as Hex };
}
